src/cpp/fastdds/publisher/PublisherImpl.cpp

```diff
@@ -257,18 +198,21 @@ DataWriter* PublisherImpl::create_datawriter(
         return nullptr;
     }
 
-    DataWriter* writer = new DataWriter(impl);
+    DataWriter* writer = new DataWriter(impl, mask);
     impl->user_datawriter_ = writer;
 
     //REGISTER THE WRITER
-    WriterQos wqos = qos.get_writerqos(qos_);
+    WriterQos wqos = qos.get_writerqos(qos_, topic->get_qos());
+    fastrtps::TopicAttributes topic_att = DataWriterImpl::get_topic_attributes(qos, *topic, type_support);
     rtps_participant_->registerWriter(impl->writer_, topic_att, wqos);
 
     {
         std::lock_guard<std::mutex> lock(mtx_writers_);
-        writers_[topic_att.getTopicName().to_string()].push_back(impl);
+        writers_[topic->get_name()].push_back(impl);
     }
 
+    topic->get_impl()->reference();
```

Comment:
Move this before creating the datawriter, or some other thread may remove the Topic meanwhile.
Best solution would be some kind of locking on the Topic, though